tools/load_balancing_tool/optimize_model.py

```diff
@@ -0,0 +1,361 @@
+                     "components.\nPerfectly scalable components would have a "
+                     "straight line. Blue 'X's designate points\nfrom data, "
+                     "red 'X's designate extrapolated data. Areas above the "
+                     "line plots represent\nthe feasible region. Global "
+                     "optimality of solution depends on the convexity of "
+                     "these line plots.\nClose graph to continue on to solve.")
+        fig.tight_layout()
+        fig.subplots_adjust(top=0.75)
+        logger.info("close graph window to continue")
+        pyplot.show()
+
+    def optimize(self):
+        """
+        Run the optimization.
+        Must set self.state using LpStatus object:
+              LpStatusOptimal    -> STATE_SOLVED_OK
+              LpStatusNotSolved  -> STATE_UNSOLVED
+              LpStatusInfeasible -> STATE_SOLVED_BAD
+              LpStatusUnbounded  -> STATE_SOLVED_BAD
+              LpStatusUndefined  -> STATE_UNDEFINED
+              -- use self.set_state(lpstatus) --
+        Returns state
+
+        If solved, then solution will be stored in self.X dictionary, indexed
+        by variable name. Suggested convention:
+        'Tice', 'Tlnd', ... for cost per component
+        'Nice', 'Nlnd', ... for ntasks per component
+        'NBice', 'NBlnd', ... for number of blocks per component
+
+        The default implementation of get_solution() returns a dictionary
+        of these variable keys and their values.
+        """
+        raise NotImplementedError
+
+    def get_solution(self):
+        """
+        Return a dictionary of the solution variables, can be overridden.
+        Default implementation returns values in self.X
+        """
+        assert self.state == self.STATE_SOLVED_OK,\
+               "solver failed, no solution available"
+        retval = {}
+        if hasattr(self,'X') and isinstance(self.X, dict):
+            for k in self.X:
+                retval[k] = self.X[k].varValue
+        return retval
+
+    def set_state(self, lpstatus):
+        if lpstatus == pulp.constants.LpStatusOptimal:
+            self.state = self.STATE_SOLVED_OK
+        elif lpstatus == pulp.constants.LpStatusNotSolved:
+            self.state = self.STATE_UNSOLVED
+        elif lpstatus == pulp.constants.LpStatusUndefined:
+            self.state = self.STATE_UNDEFINED
+        else:
+            self.state = self.STATE_SOLVED_BAD
+
+    def get_state(self):
+        return self.state
+
+    def get_state_string(self, state):
+        return self.states[state]
+
+    def get_all_variables(self):
+        return self.prob.X
+
+    def write_pe_file(self, pefilename):
+        raise NotImplementedError
+
+    def write_xml_changes(self, outfile):
+        """
+        Write out a list of xmlchange commands to implement
+        the optimal layout
+        """
+        raise NotImplementedError
+
+    def write_pe_template(self, pefilename, ntasks, nthrds, roots):
```

Comment:
Why are you reinventing the method to write env_mach_pes.xml - you should use CIME/XML/env_mach_pes.py

Reply:
Excellent questions, I have no good answers. I will attempt the suggested edits, thanks.